### PR TITLE
Containers: use untested or official container images accordingly

### PR DESCRIPTION
**Pull request by jlausuch**
There are 2 scenarios:
1) SUT=Host updates
   Here we should test the host container features and use official
   stable images for that purpose.
   These tests are triggered every day as part of maintenance updates.

2) SUT=container image updates
   Here we are targetting the images that are not released yet, so
   we use an official released HOST to test the untested images.
   For these tests, we have the variable CONTAINER_IMAGE_TYPE=untested
   These tests will be triggered whenever there is a new container image
   build in :/Update:/CR:/ToTest/images/

The PR is putting this into practice and using the right arrays depending
on the situation.

[type description here, PLEASE, REMOVE THIS LINE, PLACEHOLDER, BEFORE SUBMITTING THIS PULL REQUEST]

- Related ticket: https://progress.opensuse.org/issues/93480

### CODE REVIEW

File: lib/containers/urls.pm [hunk 1]
```diff
@@ -57,52 +57,46 @@ sub get_suse_container_urls {
...
+            push @official_images, "registry.suse.com/suse/sles${nodashversion}";
         }
     }
     elsif (is_sle(">=15", $version) && is_released) {
         my $lowerversion = lc $version;
         # Location for maintenance builds
-        push @image_names,  "registry.suse.de/suse/sle-${lowerversion}/update/cr/totest/images/suse/sle15:${dotversion}";
-        push @stable_names, "registry.suse.com/suse/sle15:${dotversion}";
-    }
-    elsif (is_sle(">=15-sp3", $version)) {
```
Comment (pdostal) [inline lib/containers/urls.pm:77]: Why is this deleted?

File: tests/containers/rootless_podman.pm [hunk 2]
```diff
@@ -30,7 +30,7 @@ use version_utils 'is_sle';
 
 sub run {
     my ($self) = @_;
-    my ($image_names, $stable_names) = get_suse_container_urls();
+    my ($untested_images, $official_images) = get_suse_container_urls($version);
```
Comment (pdostal) [inline tests/containers/rootless_podman.pm:33]: There is no `$version` defined.
```suggestion
    my ($untested_images, $official_images) = get_suse_container_urls();
```

**Review by pdostal — APPROVED** [review]
Fine by me, I will double check once more later. We for sure need verification runs.

Just please documment the `CONTAINER_IMAGE_TYPE` variable to `variables.md` as well.

File: lib/containers/urls.pm (hunk 1 above)
Comment (jlausuch) [reply lib/containers/urls.pm:77]: 15-SP3 is already released. We should enable 15-SP3 in the bot as well, so we should pull also those builds form the same place as the others.

Comment (jlausuch) [reply lib/containers/urls.pm:77]: > Fine by me, I will double check once more later. We for sure need verification runs.
> 
> Just please documment the `CONTAINER_IMAGE_TYPE` variable to `variables.md` as well.

I actually was thinking of using it a bit differently... Something like `CONTAINERS_UNTESTED_IMAGES=1` instead of depending on a string... WDYT?

Comment (jlausuch) [reply lib/containers/urls.pm:77]: Just changed that var. The OpenQA bot is also configured to push `CONTAINERS_UNTESTED_IMAGES=1`

Comment (Vogtinator) [reply lib/containers/urls.pm:77]: Instead of deleting, just bump to SP4?

File: variables.md [hunk 3]
```diff
@@ -28,6 +28,7 @@ CHECK_RELEASENOTES | boolean | false | Loads `installation/releasenotes` test mo
 CHECK_RELEASENOTES_ORIGIN | boolean | false | Loads `installation/releasenotes_origin` test module.
 CHECKSUM_* | string | | SHA256 checksum of the * medium. E.g. CHECKSUM_ISO_1 for ISO_1.
 CHECKSUM_FAILED | string | | Variable is set if checksum of installation medium fails to visualize error in the test module and not just put this information in the autoinst log file.
+CONTAINERS_UNTESTED_IMAGES | boolean | false | Used by container tests to decide if pulling an unreleased image from registry.suse.de or a released image from registry.suse.com.
```
Comment (Vogtinator) [inline variables.md:31]: SLE specific.
```suggestion
CONTAINERS_UNTESTED_IMAGES | boolean | false | Whether to use `untested_images` or `official_images` from `lib/containers/urls.pm`.
```

File: lib/containers/urls.pm [hunk 4]
```diff
@@ -57,52 +57,46 @@ sub get_suse_container_urls {
     my $dotversion = $version =~ s/-SP/./r;                    # 15 -> 15, 15-SP1 -> 15.1
     $dotversion = "${dotversion}.0" if $dotversion !~ /\./;    # 15 -> 15.0
 
-    my @image_names  = ();
-    my @stable_names = ();
+    my @untested_images = ();
+    my @official_images = ();
```
Comment (Vogtinator) [inline lib/containers/urls.pm:61]: The untested images are also official

```suggestion
    my @released_images = ();
```

**Comment by Vogtinator** [comment]
Previously, the `docker_image` and `podman_image` tests had the goal of testing the unreleased images and the `docker`/`podman`/`docker_buildah` tests had the container runtime as focus. This is to be able to distinguish between container image and runtime failures.

With this PR, the ability to do that is lost, and it just duplicates some of the testing. And for some of the runtime tests it actually can't use the unreleased image yet, because it has mismatched repos. For instance `basic_container_tests` does `podman container exec tw zypper -n in curl` but doing that with an unreleased image results in a downgrade and will most likely fail.

**Comment by pdostal** [comment]
> Previously, the `docker_image` and `podman_image` tests had the goal of testing the unreleased images and the `docker`/`podman`/`docker_buildah` tests had the container runtime as focus. This is to be able to distinguish between container image and runtime failures.

No, in my understanding nothing changed: The `docker` / `podman` are runtime tests with some 'generic' images. Then the `docker_image` / `podman_image` are tests of our based images.

**Comment by jlausuch** [comment]
> > > Previously, the `docker_image` and `podman_image` tests had the goal of testing the unreleased images and the `docker`/`podman`/`docker_buildah` tests had the container runtime as focus. This is to be able to distinguish between container image and runtime failures.
> > 
> > 
> > No, in my understanding nothing changed: The `docker` / `podman` are runtime tests with some 'generic' images. Then the `docker_image` / `podman_image` are tests of our based images.
> 
> Indeed - I thought this changed `docker`/`podman` as well, but it doesn't.
> 
> I'm not sure about `buildah_docker` and `buildah_podman`. (I noticed that those modules are missing from `jeos-containers`, is that intentional?) The test currently runs `zypper -n in python3 python3-pip curl` in the untested image, but it seems like this was broken even before this PR though.

Those 2 tests will be part of container runtime testing, and they will pull the released images. And thanks for the heads up about Jeos. I plan to split the jeos container tests as well and test only runtime with Released images. But that's for another PR :) 

If you agree I will run a bunch of VRs in OSD and O3 with affected tests by this code.

**Comment by Vogtinator** [comment]
> > > > Previously, the `docker_image` and `podman_image` tests had the goal of testing the unreleased images and the `docker`/`podman`/`docker_buildah` tests had the container runtime as focus. This is to be able to distinguish between container image and runtime failures.
> > > 
> > > 
> > > No, in my understanding nothing changed: The `docker` / `podman` are runtime tests with some 'generic' images. Then the `docker_image` / `podman_image` are tests of our based images.
> > 
> > 
> > Indeed - I thought this changed `docker`/`podman` as well, but it doesn't.
> > I'm not sure about `buildah_docker` and `buildah_podman`. (I noticed that those modules are missing from `jeos-containers`, is that intentional?) The test currently runs `zypper -n in python3 python3-pip curl` in the untested image, but it seems like this was broken even before this PR though.
> 
> Those 2 tests will be part of container runtime testing, and they will pull the released images. And thanks for the heads up about Jeos. I plan to split the jeos container tests as well and test only runtime with Released images. But that's for another PR :)

What's the benefit of splitting the JeOS container tests into separate jobs?
 
> If you agree I will run a bunch of VRs in OSD and O3 with affected tests by this code.

Sounds good!

**Comment by jlausuch** [comment]
> 
> What's the benefit of splitting the JeOS container tests into separate jobs?
It's not clear to me how, so maybe you have good insight here. Today we are testing docker_image and podman_image in JeOS as well with untested images. So, I guess we should either 1) skip those tests 2) run them but with released images, which is also fine.

> 
> > If you agree I will run a bunch of VRs in OSD and O3 with affected tests by this code.
> 
> Sounds good!

**Comment by jlausuch** [comment]
openSUSE: 
- [TW extra_tests_textmode_containers](https://openqa.opensuse.org/tests/1787112)
- [TW JeOS jeos-containers](https://openqa.opensuse.org/tests/1787113)
- [Leap 15.3 ](https://openqa.opensuse.org/tests/1787119): failure not related to PR (maybe bug?)

SLE Host Updates:
TBD

SLE Image Updates:
- [sle-15-SP3_image_on_sle-15-SP3_host](http://fromm.arch.suse.de/tests/1625) [bsc#1187176](https://bugzilla.suse.com/show_bug.cgi?id=1187176)
- [sle-15-SP3_image_on_sle-15-SP2_host](http://fromm.arch.suse.de/tests/1624)
- [sle-15-SP3_image_on_sle-15-SP1_host](http://fromm.arch.suse.de/tests/1623)
- [sle-15-SP3_image_on_sle-15_host](http://fromm.arch.suse.de/tests/1626)
- [sle-15-SP3_image_on_sle-12-SP5_host](http://fromm.arch.suse.de/tests/1622)

File: tests/containers/buildah_podman.pm [hunk 5]
```diff
@@ -24,29 +24,34 @@ use containers::urls 'get_suse_container_urls';
...
+                my ($prefix_img_name) = $iname =~ /([^\/:]+)(:.+)?$/;
+                record_info $prefix_img_name;
+                test_opensuse_based_image(image => "${prefix_img_name}-working-container", runtime => 'buildah');
+                # Due to the steps from the test_opensuse_based_image previously,
+                # the image has been committed as refreshed
+                test_containered_app(runtime => 'podman',
+                    buildah    => 1,
+                    dockerfile => 'Dockerfile.suse',
+                    base       => 'refreshed');
+            }
```
Comment (okurz) [inline tests/containers/buildah_podman.pm:54]: This looks like a near-duplicate to the above. Looks like you are just fooling tools/detect_code_dups with the line `record_info $prefix_img_name;` in the middle. I strongly suggest to resolve the duplication first

File: tests/containers/buildah_docker.pm [hunk 6]
```diff
@@ -22,30 +22,35 @@ use containers::common;
...
-            my ($prefix_img_name) = $iname =~ /([^\/:]+)(:.+)?$/;
-            test_opensuse_based_image(image => "${prefix_img_name}-working-container", runtime => 'buildah');
-            # Due to the steps from the test_opensuse_based_image previously,
-            # the image has been committed as refreshed
-            test_containered_app(runtime => 'docker',
-                buildah    => 1,
-                dockerfile => 'Dockerfile.suse',
-                base       => 'refreshed');
+    # We may test either one specific image VERSION or comma-separated CONTAINER_IMAGES
+    my $versions = get_var('CONTAINER_IMAGE_VERSIONS', get_required_var('VERSION'));
```
Comment (asmorodskyi) [inline tests/containers/buildah_docker.pm:35]: this loop looks exactly the same as the one in `buildah_podman` . maybe we need to extract it to a function ? ( I would accept if you will say that it can be done in next PR :) )

**Review by asmorodskyi — APPROVED** [review]
LGTM

File: tests/containers/buildah_podman.pm (hunk 5 above)
Comment (asmorodskyi) [reply tests/containers/buildah_podman.pm:54]: funny that we did basically same comment at the same time :) but you wrote yours in more demanding way. And obviously it is code duplication . But I always have strong opinion - I can not **require** from the person to fix **all** the issues which exist in the file which he/she modified. You can only require to not introduce new. If you look careful in the diff you will see that this duplication was there before this patch. So it would be nice if Jose will fix it but it is for not a strong requirement for this PR to be merged .

Comment (jlausuch) [reply tests/containers/buildah_podman.pm]: removed

File: tests/containers/buildah_docker.pm (hunk 6 above)
Comment (jlausuch) [reply tests/containers/buildah_docker.pm:35]: Yes, I would like to leave it for another PR. Actually `buildah_podman.pm` and `buildah_docker.pm` have the exact same code but different runtime... So I guess that deserves some small refactor and maybe combine into 1 `.pm` with some input variable.. But that's not the purpose of this PR, so as you said, next one :)

File: tests/containers/buildah_podman.pm (hunk 5 above)
Comment (okurz) [reply tests/containers/buildah_podman.pm]: @asmorodskyi Correct. I should have written that I already saw the duplication in the old code

Comment (jlausuch) [reply tests/containers/buildah_podman.pm]: [poo#94000](https://progress.opensuse.org/issues/94000)

Comment (okurz) [reply tests/containers/buildah_podman.pm]: cool, thx. solved :)

**Comment by Vogtinator** [comment]
> openSUSE:
> 
> * [TW extra_tests_textmode_containers](https://openqa.opensuse.org/tests/1787112)
> * [TW JeOS jeos-containers](https://openqa.opensuse.org/tests/1787113)

Appears to test the wrong image in `podman_image`, the released one: https://openqa.opensuse.org/tests/1787113#step/podman_image/27

> * [Leap 15.3 ](https://openqa.opensuse.org/tests/1787119): failure not related to PR (maybe bug?)

Apparently the docker update got released, so https://github.com/jlausuch/os-autoinst-distri-opensuse/commit/03551bba22ddc6ba6d8e42ee69f2ccbe1ec7a226#diff-3b34101ea0df2e75ea46097ce9ac2b57880f89129ee82472f20b3e689d6e4132 needs to account for that as well.

> SLE Host Updates:
> [15-SP3 sle_image_on_sle_host](http://fromm.arch.suse.de/tests/1629#step/docker_image/62) (image from registry.suse.com) -> need to fix schedule to execute also podman tests, but different PR.
> [15-SP3 engines_and_tools](http://fromm.arch.suse.de/tests/1628) [poo#93889](https://progress.opensuse.org/issues/93889)
> 
> SLE Image Updates:
> 
> * [sle-15-SP3_image_on_sle-15-SP3_host](http://fromm.arch.suse.de/tests/1625) [bsc#1187176](https://bugzilla.suse.com/show_bug.cgi?id=1187176)
> * [sle-15-SP3_image_on_sle-15-SP2_host](http://fromm.arch.suse.de/tests/1624)
> * [sle-15-SP3_image_on_sle-15-SP1_host](http://fromm.arch.suse.de/tests/1623)
> * [sle-15-SP3_image_on_sle-15_host](http://fromm.arch.suse.de/tests/1626)
> * [sle-15-SP3_image_on_sle-12-SP5_host](http://fromm.arch.suse.de/tests/1622)

SLE results LGTM

**Comment by jlausuch** [comment]
> > openSUSE:
> > 
> > * [TW extra_tests_textmode_containers](https://openqa.opensuse.org/tests/1787112)
> > * [TW JeOS jeos-containers](https://openqa.opensuse.org/tests/1787113)
> 
> Appears to test the wrong image in `podman_image`, the released one: https://openqa.opensuse.org/tests/1787113#step/podman_image/27

Well, we are testing here the Host, so we should use released images for that. The fact is that we should also have jobs for testing unreleased opensuse images as I am doing for SLE. Do you agree to keep it as it is and then we create jobs for new image builds testing unreleased images?

> 
> > * [Leap 15.3 ](https://openqa.opensuse.org/tests/1787119): failure not related to PR (maybe bug?)
> 
> Apparently the docker update got released, so [jlausuch@03551bb#diff-3b34101ea0df2e75ea46097ce9ac2b57880f89129ee82472f20b3e689d6e4132](https://github.com/jlausuch/os-autoinst-distri-opensuse/commit/03551bba22ddc6ba6d8e42ee69f2ccbe1ec7a226#diff-3b34101ea0df2e75ea46097ce9ac2b57880f89129ee82472f20b3e689d6e4132) needs to account for that as well.

Ticket created for this [poo#94018](https://progress.opensuse.org/issues/94018). Gonna fix it in separate PR.

**Comment by Vogtinator** [comment]
> > > openSUSE:
> > > 
> > > * [TW extra_tests_textmode_containers](https://openqa.opensuse.org/tests/1787112)
> > > * [TW JeOS jeos-containers](https://openqa.opensuse.org/tests/1787113)
> > 
> > 
> > Appears to test the wrong image in `podman_image`, the released one: https://openqa.opensuse.org/tests/1787113#step/podman_image/27
> 
> Well, we are testing here the Host, so we should use released images for that. The fact is that we should also have jobs for testing unreleased opensuse images as I am doing for SLE. Do you agree to keep it as it is and then we create jobs for new image builds testing unreleased images?

Yes, but please do VRs for those as well.

> > > * [Leap 15.3 ](https://openqa.opensuse.org/tests/1787119): failure not related to PR (maybe bug?)
> > 
> > Apparently the docker update got released, so [jlausuch@03551bb#diff-3b34101ea0df2e75ea46097ce9ac2b57880f89129ee82472f20b3e689d6e4132](https://github.com/jlausuch/os-autoinst-distri-opensuse/commit/03551bba22ddc6ba6d8e42ee69f2ccbe1ec7a226#diff-3b34101ea0df2e75ea46097ce9ac2b57880f89129ee82472f20b3e689d6e4132) needs to account for that as well.
> 
> Ticket created for this [poo#94018](https://progress.opensuse.org/issues/94018). Gonna fix it in separate PR.

**Comment by jlausuch** [comment]
> > > > openSUSE:
> > > > 
> > > > * [TW extra_tests_textmode_containers](https://openqa.opensuse.org/tests/1787112)
> > > > * [TW JeOS jeos-containers](https://openqa.opensuse.org/tests/1787113)
> > > 
> > > 
> > > Appears to test the wrong image in `podman_image`, the released one: https://openqa.opensuse.org/tests/1787113#step/podman_image/27
> > 
> > 
> > Well, we are testing here the Host, so we should use released images for that. The fact is that we should also have jobs for testing unreleased opensuse images as I am doing for SLE. Do you agree to keep it as it is and then we create jobs for new image builds testing unreleased images?
> 
> Yes, but please do VRs for those as well.

The jobs are not in place yet, that will take some time to setup.

**Comment by jlausuch** [comment]
VRs for TW: 

- [TW extra_tests_textmode_containers](http://fromm.arch.suse.de/tests/1635#step/podman_image/27)
- [TW JeOS containers](http://fromm.arch.suse.de/tests/1634#step/podman_image/27)
- [TW on CentOS](http://fromm.arch.suse.de/tests/1637#step/podman_image/27)
- [TW on Ubuntu](http://fromm.arch.suse.de/tests/1638#step/podman_image/47)
